packages/zero-client/src/client/mutator-proxy.test.ts

```diff
@@ -0,0 +1,736 @@
+      const proxy = new MutatorProxy(
+        manager,
+        mutationTracker,
+        onApplicationError,
+      );
+
+      const thrownError = new Error('mutator failed');
+      const mutator = vi.fn(() => {
+        throw thrownError;
+      });
+      const wrapped = proxy.wrapCustomMutator(mutator);
+
+      const result = wrapped();
+
+      const clientResult = await result.client;
+      const serverResult = await result.server;
+
+      expect(clientResult.type).toBe('error');
+      expect(clientResult).toEqual({
+        type: 'error',
+        error: {
+          type: 'app',
+          message: 'mutator failed',
+          details: undefined,
+        },
+      });
+
+      expect(serverResult).toEqual(clientResult);
+      expect(mutator).toHaveBeenCalledTimes(1);
+
+      // onApplicationError should be called once
+      expect(onApplicationError).toHaveBeenCalledTimes(1);
+      expect(onApplicationError).toHaveBeenCalledWith(
+        expect.objectContaining({
+          message: 'mutator failed',
+        }),
+      );
+      expect(isApplicationError(onApplicationError.mock.calls[0][0])).toBe(
+        true,
+      );
+    });
+
+    test('returns success when mutator succeeds', async () => {
+      const {manager, mutationTracker} = createMockConnectionManager();
+      const onApplicationError = vi.fn();
+      const proxy = new MutatorProxy(
+        manager,
+        mutationTracker,
+        onApplicationError,
+      );
+
+      const mutator = vi.fn(() => ({
+        client: Promise.resolve(),
+        server: Promise.resolve(),
+      }));
+      const wrapped = proxy.wrapCustomMutator(mutator);
+
+      const result = wrapped();
+
+      const clientResult = await result.client;
+      const serverResult = await result.server;
+
+      expect(clientResult).toEqual({type: 'success'});
+      expect(serverResult).toEqual({type: 'success'});
+      expect(mutator).toHaveBeenCalledTimes(1);
+
+      // onApplicationError should not be called on success
+      expect(onApplicationError).not.toHaveBeenCalled();
+    });
+
+    test('wraps client promise rejection as app error', async () => {
+      const {manager, mutationTracker} = createMockConnectionManager();
+      const onApplicationError = vi.fn();
+      const proxy = new MutatorProxy(
+        manager,
+        mutationTracker,
+        onApplicationError,
+      );
+
+      const clientError = new Error('client failed');
+      const mutator = vi.fn(() => ({
+        client: Promise.reject(clientError),
+        server: Promise.resolve(),
+      }));
+      const wrapped = proxy.wrapCustomMutator(mutator);
+
+      const result = wrapped();
+
+      const clientResult = await result.client;
+      const serverResult = await result.server;
+
+      expect(clientResult.type).toBe('error');
+      expect(clientResult).toEqual({
+        type: 'error',
+        error: {
+          type: 'app',
+          message: 'client failed',
```

Comment:
Did we lose the underlying Error object here? How do I get the stack trace for example?